ensembleperturbation/perturbation/atcf.py

```diff
@@ -601,56 +727,198 @@ def __init__(self):
+        # get Vr for the X-kt isotach in this quadrant
+        Vr = (
+            dataframe['isotach_radius'].values
+            * MaximumSustainedWindSpeed.unit  # LC12 factor and 20-deg CCW rotation
+            - ts_factor(
+                dataframe[self.name].values, dataframe[RadiusOfMaximumWinds.name].values
+            )
+            * translation_speed
+            * numpy.sin(numpy.deg2rad(self.quadrant_angle + 20))
+        ) / BLAdj
+        # get Rmax with units
+        Rmax = dataframe[RadiusOfMaximumWinds.name].values * RadiusOfMaximumWinds.unit
+        # get Coriolis and compute Rossby number (inverse of)
+        f = 2 * omega * numpy.sin(numpy.deg2rad(dataframe['latitude'].values))
+        Ro_inv = (Rmax * f / Vmax).to('dimensionless')
+        # get central pessure and compute standard Holland B parameter as initial guess
+        DelP = dataframe[BackgroundPressure.name] - dataframe[CentralPressure.name]
+        DelP = DelP.values * CentralPressure.unit
+        B = Vmax.to('m/s') ** 2 * AIR_DENSITY * E1 / DelP.to('Pa')
+        # now found the Bg and phi (and subsequent B) that actually fits
+        # through both the isotach and the Rmax value
+        return Vmax, Vr, Rmax, Ro_inv, f, B
+
+    def perturb(
+        self, vortex_dataframe: DataFrame, original_dataframe, inplace: bool = False,
+    ) -> DataFrame:
+        """ Compute new isotach radius based on Holland profile using perturbed Rmax/Vmax/speed values"""
+        # retrieve the Vmax, Vr(isotach) at the top of the boundary layer, and Rmax
+        # in the original and perturbed dataframe
+        Vmax_old, Vr_old, Rmax_old, Roinv_old, f_old, B_ini = self.get_isotach_properties(
+            dataframe=original_dataframe
+        )
+        Vmax_new, Vr_new, Rmax_new, Roinv_new, f_new, _ = self.get_isotach_properties(
+            dataframe=vortex_dataframe
+        )
+        # determine original isotach_rad and Rrat
+        isotach_rad = original_dataframe[self.name].values * RadiusOfMaximumWinds.unit
+        isotach_rad[isotach_rad == 0] = numpy.nan
+        Rrat = Rmax_old / isotach_rad  # [nmi/nmi] = []
+        Rrat[abs(Rrat - 1.0) < 1e-3] = 0.999  # ensure not exactly 1
+        # find B from original velocity profile using GAHM (traditional Holland B as first guess)
+        B = self.find_parameter_from_GAHM_profile(
+            Vr_old, Vmax_old, f_old, Roinv_old, B=B_ini, isotach_rad=isotach_rad, Rrat=Rrat
+        )
+        # correct where B is nan
+        invalid = numpy.isnan(B)
+        B[invalid] = B_ini[invalid]
+        # preserving B, find new GAHM Bg and phi using new Ro_inv
+        Bg, phi = self.find_GAHM_parameters(B, Roinv_new)
+        # determine guess of new Rrat
+        Rrat = Rmax_new / isotach_rad
+        # correct where Rrat is nan
+        invalid = numpy.isnan(Rrat)
+        Rrat[invalid] = Vr_new[invalid] / Vmax_new[invalid]
```

Comment:
If isotach_rad is zero (NaN) then the guess for Rrat will be NaN so this correction allows for a non-zero guess and non-zero solution to be potentially be found for the pertubed isotach radius.